D3b 1413 add specimen transformers

calkinsh opened the pull request:
Add specimen transformers and tests

Once we can transform nested lists we will need to add:

Processing.procedure.coding
Processing.additive
Container.identifier
Container.type.coding
Specimen.additiveCodeableConcept.coding

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen.py:
```diff
@@ -0,0 +1,209 @@
+                "type": "str",
+            },
+            "accession_identifier_value": {"fhir_key": "value", "type": "str"},
+            "accession_identifier_period": {
+                "fhir_key": "period",
+                "type": "str",
+            },
```
According to the specification here (https://hl7.org/fhir/R4/datatypes.html#Period), I believe the structure should look like this:

```python
{
    "accession_identifier_period_start": {
        "fhir_key": "period.start",
        "type": "datetime",
    },
    "accession_identifier_period_end": {
        "fhir_key": "period.end",
        "type": "datetime",
    }
}
```

Let me know if you see it differently!

calkinsh replied:
yes you're right! nice catch, thank you. resolved in new commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen.py:
```diff
@@ -0,0 +1,209 @@
+                "fhir_key": "period",
+                "type": "str",
+            },
+            "accession_identifier_assigner_reference": {
+                "fhir_key": "assigner.reference",
+                "type": "str",
+            },
+            "accession_identifier_assigner_display": {
+                "fhir_key": "assigner.display",
+                "type": "str",
+            },
```
Since we're aiming to curate the resource type as described in [the FHIR specification](https://hl7.org/fhir/R4/references.html#Reference), do you think we should also include the following field?

```python
{
    "accession_identifier_assigner_type": {
        "fhir_key": "assigner.type",
        "type": "str",
    }
}
```

calkinsh replied:
resolved in commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen.py:
```diff
@@ -0,0 +1,209 @@
+                "fhir_key": "reference",
+                "type": "str",
+            },
+            "subject_display": {"fhir_key": "display", "type": "str"},
```
```python
{
    "subject_type": {
        "fhir_key": "type",
        "type": "str",
    }
}
```

calkinsh replied:
resolved in new commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen.py:
```diff
@@ -0,0 +1,209 @@
+        "fhir_path": "collection.collector",
+        "fhir_reference": "collection_collector_reference",
+        "columns": {
+            "collection_collector_reference": {
+                "fhir_key": "reference",
+                "type": "str",
+            },
+            "collection_collector_display": {
+                "fhir_key": "display",
+                "type": "str",
```
```python
{
    "collection_collector_type": {
        "fhir_key": "type",
        "type": "str",
    }
}
```

calkinsh replied:
resolved in new commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen.py:
```diff
@@ -0,0 +1,209 @@
+        },
+    },
+    {
+        "fhir_path": "collection.Duration",
```
`collection.duration`

calkinsh replied:
resolved in new commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_collection_body_site_coding.py:
```diff
@@ -0,0 +1,66 @@
+
+    Attributes:
+        resource_type (str): The type of FHIR resource being transformed ('Specimen').
+        subtype (str): Specifies the sub-element of the resource to focus on ('collection.bodySite.coding').
```
`collection_body_site_coding`

calkinsh replied:
@liberaliscomputing are you suggesting that instead of saying 'collection.bodySite.coding' here this should say collection_body_site_coding?  I dont think I agree - in the examples charles provided he using json pointer type notation (see https://github.com/radiant-network/radiant-fhir-transformers/blob/main/radiant_fhir_transform_cli/transform/classes/observation/observation_category_coding.py as an example) and not the expanded notation that uses underscores.

liberaliscomputing replied:
This isn’t critical to functionality, so feel free to disregard if you prefer. But I wanted to share a few thoughts on docstring style for consistency:

- According to [PEP 257](https://peps.python.org/pep-0257/), while it doesn’t explicitly mandate how to name attributes in docstrings, it does emphasize that class docstrings should describe the public methods and instance variables. That suggests we should use **snake\_case** in the `Attributes` and `Methods` sections to reflect actual code usage. Dot notation makes more sense when referring to external data structures, not internal APIs.
- There's currently some inconsistency in the repo:
    - Uses dot notation: [observation_category_coding.py](https://github.com/radiant-network/radiant-fhir-transformers/blob/main/radiant_fhir_transform_cli/transform/classes/observation/observation_category_coding.py)
    - Uses snake\_case: [observation_code_coding.py](https://github.com/radiant-network/radiant-fhir-transformers/blob/main/radiant_fhir_transform_cli/transform/classes/observation/observation_code_coding.py)

Ultimately, I’m totally fine either way as long as it's clear to users—it’s more about aiming for consistency across transformer classes.

calkinsh replied:
I think I am in favor of dot notation here because it makes it very clear where, inside the JSON object, the transformer is focused.  If we list it out like collection_body_site_coding it is not clear to the person reading it where each element starts. The underscores replace dots but they ALSO separate words that have no dots, so once we switch to snake case I feel we are more ambiguous about what exactly is being transformed.  For example, someone could reasonably think that collection_body_site_coding would translate to a JSON pointer of collection.body.site.coding... which is incorrect.  So ultimately while i agree with the best practice if all things are neutral, i disagree in this case since I think it adds some information loss to the process which is not ideal.  Thoughts?

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_collection_body_site_coding.py:
```diff
@@ -0,0 +1,66 @@
+    Methods:
+        __init__():
+            Initializes the SpecimenBodySiteTransformer instance with the resource type 'Specimen',
+            subtype 'collection.bodySite.coding', and the specified transformation dictionary.
```
`collection_body_site_coding`

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_collection_fasting_status_codeable_concept_coding.py:
```diff
@@ -0,0 +1,70 @@
+
+    Attributes:
+        resource_type (str): The type of FHIR resource being transformed ('Specimen').
+        subtype (str): Specifies the sub-element of the resource to focus on ('collection.fastingStatusCodeableConcept.coding').
```
`collection_fasting_status_codeable_concept_coding`

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_collection_fasting_status_codeable_concept_coding.py:
```diff
@@ -0,0 +1,70 @@
+    Methods:
+        __init__():
+            Initializes the SpecimenfastingStatusCodeableConceptCodingTransformer instance with the resource type 'Specimen',
+            subtype 'collection.fastingStatusCodeableConcept.coding', and the specified transformation dictionary.
```
`collection_fasting_status_codeable_concept_coding`

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_collection_method_coding.py:
```diff
@@ -0,0 +1,65 @@
+
+    Attributes:
+        resource_type (str): The type of FHIR resource being transformed ('Specimen').
+        subtype (str): Specifies the sub-element of the resource to focus on ('collection.method.coding').
```
`collection_method_coding`

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_collection_method_coding.py:
```diff
@@ -0,0 +1,65 @@
+    Methods:
+        __init__():
+            Initializes the SpecimenCollectionMethodCodingTransformer instance with the resource type 'Specimen',
+            subtype 'collection.method.coding', and the specified transformation dictionary.
```
`collection_method_coding`

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_condition.py:
```diff
@@ -0,0 +1,53 @@
+    Transformer class for the 'Specimen' resource in FHIR, focusing on the 'condition' element.
+
+    This class transforms FHIR Specimen JSON objects into flat dictionaries suitable for CSV output,
+    extracting and processing information from the 'category' field.
```
`condition`

calkinsh replied:
resolved in new commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_container.py:
```diff
@@ -0,0 +1,104 @@
+                "fhir_key": "specimen.additiveCodeableConcept.coding",
+                "type": "str",
+            },
+            "container_specimen_additive_reference_reference": {
+                "fhir_key": "specimen.additiveReference.reference",
+                "type": "str",
+            },
+            "container_specimen_additive_reference_display": {
+                "fhir_key": "specimen.additiveReference.display",
+                "type": "str",
```
```python
{
    "container_specimen_additive_type": {
        "fhir_key": "specimen.additiveReference.type",
        "type": "str",
    }
}
```

calkinsh replied:
resolved in new commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_note.py:
```diff
@@ -0,0 +1,63 @@
+        "columns": {
+            "note_text": {"fhir_key": "text", "type": "str"},
+            "note_author_string": {"fhir_key": "authorString", "type": "str"},
+            "note_author_reference_reference": {
+                "fhir_key": "authorReference.reference",
+                "type": "str",
+            },
+            "note_author_reference_display": {
+                "fhir_key": "authorReference.display",
+                "type": "str",
+            },
```
```python
{
    "note_author_reference_type": {
        "fhir_key": "authorReference.type",
        "type": "str",
    }
}
```

calkinsh replied:
resolved in new commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_parent.py:
```diff
@@ -0,0 +1,60 @@
+        "fhir_path": "parent",
+        "fhir_reference": "parent_reference",
+        "columns": {
+            "parent_reference": {
+                "fhir_key": "reference",
+                "type": "str",
+            },
+            "parent_display": {
+                "fhir_key": "display",
+                "type": "str",
+            },
```
```python
{
    "parent_type": {
        "fhir_key": "type",
        "type": "str",
    }
}
```

calkinsh replied:
resolved in commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_parent.py:
```diff
@@ -0,0 +1,60 @@
+    Transformer class for the 'Specimen' resource in FHIR, focusing on the 'parent' element.
+
+    This class transforms FHIR Specimen JSON objects into flat dictionaries suitable for CSV output,
+    extracting and processing information from the 'extension' field.
```
`parent`

calkinsh replied:
resolved in new commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_request.py:
```diff
@@ -0,0 +1,60 @@
+        "fhir_path": "request",
+        "fhir_reference": "request_reference",
+        "columns": {
+            "request_reference": {
+                "fhir_key": "reference",
+                "type": "str",
+            },
+            "request_display": {
+                "fhir_key": "display",
+                "type": "str",
+            },
```
```python
{
    "request_type": {
        "fhir_key": "type",
        "type": "str",
    }
}
```

calkinsh replied:
resolved in new commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_type_coding.py:
```diff
@@ -0,0 +1,54 @@
+    Transformer class for the 'Specimen' resource in FHIR, focusing on the 'type.coding' element.
+
+    This class transforms FHIR Specimen JSON objects into flat dictionaries suitable for CSV output,
+    extracting and processing information from the 'category.coding' field.
```
`type.coding`

calkinsh replied:
resolved in commit

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_type_coding.py:
```diff
@@ -0,0 +1,54 @@
+
+    Attributes:
+        resource_type (str): The type of FHIR resource being transformed ('Specimen').
+        subtype (str): Specifies the sub-element of the resource to focus on ('type.coding').
```
`type_coding`

liberaliscomputing commented on radiant_fhir_transform_cli/transform/classes/specimen/specimen_type_coding.py:
```diff
@@ -0,0 +1,54 @@
+    Methods:
+        __init__():
+            Initializes the SpecimenTypeCodingTransformer instance with the resource type 'Specimen',
+            subtype 'type.coding', and the specified transformation dictionary.
```
`type_coding`

liberaliscomputing commented on tests/data/specimen/specimen.py:
```diff
@@ -0,0 +1,80 @@
+        "accession_identifier_system": "http://lab.acme.org/specimens/2011",
+        "accession_identifier_value": "X352356",
+        "accession_identifier_period": None,
+        "accession_identifier_assigner_reference": None,
+        "accession_identifier_assigner_display": None,
```
`"accession_identifier_assigner_type": None,`

calkinsh replied:
resolved in new commit

liberaliscomputing commented on tests/data/specimen/specimen.py:
```diff
@@ -0,0 +1,80 @@
+        "accession_identifier_assigner_reference": None,
+        "accession_identifier_assigner_display": None,
+        "type_text": "venous blood",
+        "subject_reference": "example",
+        "subject_display": "Peter Patient",
```
`"subject_type": "Patient",`

calkinsh replied:
resolved in new commit

liberaliscomputing commented on tests/data/specimen/specimen.py:
```diff
@@ -0,0 +1,80 @@
+        "subject_reference": "example",
+        "subject_display": "Peter Patient",
+        "received_time": "2011-03-04T07:03:00Z",
+        "collection_collector_reference": "example",
+        "collection_collector_display": None,
```
`"collection_collector_type": "Practitioner",`

calkinsh replied:
resolved in new commit

liberaliscomputing commented on tests/data/specimen/specimen_note.py:
```diff
@@ -0,0 +1,51 @@
+        "specimen_id": "101",
+        "note_text": "Specimen is grossly lipemic",
+        "note_author_string": "Doctor Judy",
+        "note_author_reference_reference": None,
+        "note_author_reference_display": None,
```
`"note_author_reference_type": None,`

calkinsh replied:
resolved in new commit

liberaliscomputing commented on tests/data/specimen/specimen_parent.py:
```diff
@@ -0,0 +1,48 @@
+EXPECTED_OUTPUT = [
+    {
+        "specimen_id": "101",
+        "parent_reference": "example1",
+        "parent_display": "Whole Blood",
```
`"parent_type": "Specimen",`

liberaliscomputing commented on tests/data/specimen/specimen_request.py:
```diff
@@ -0,0 +1,48 @@
+EXPECTED_OUTPUT = [
+    {
+        "specimen_id": "101",
+        "request_reference": "example",
+        "request_display": None,
```
`"request_type": "ServiceRequest",`

calkinsh replied:
resolved in new commit

calkinsh commented:
@liberaliscomputing i think i addressed all your comments except for those that i'm not sure were appropriate. Happy to discuss further but can you take another look?
